lib/tools/parse-id3.js

```diff
@@ -0,0 +1,250 @@
+
+      if (descriptionEndIndex === -1) {
+        return;
+      }
+
+      // parse the text fields
+      frame.description = parseUtf8(frame.data, 1, descriptionEndIndex);
+      // do not include the null terminator in the tag value
+      // frames that allow different types of encoding contain terminated text
+      // [ID3v2.4.0 section 4.]
+      frame.value = parseUtf8(
+        frame.data,
+        descriptionEndIndex + 1,
+        frame.data.length
+      ).replace(/\0*$/, '');
+      frame.data = frame.value;
+    },
+    'W*': function(frame) {
+      // parse URL field; URL fields are always represented as ISO-8859-1 [ID3v2.4.0 section 4.]
+      // if the value is followed by a string termination all the following information should be ignored [ID3v2.4.0 section 4.3]
+      frame.url = parseIso88591(frame.data, 0, frame.data.length).replace(/\0.*$/, '');
+    },
+    'WXXX': function(frame) {
+      var descriptionEndIndex;
+
+      if (frame.data[0] !== textEncodingDescriptionByte.Utf8) {
+        // ignore frames with unrecognized character encodings
+        return;
+      }
+
+      descriptionEndIndex = typedArrayIndexOf(frame.data, 0, 1);
+
+      if (descriptionEndIndex === -1) {
+        return;
+      }
+
+      // parse the description and URL fields
+      frame.description = parseUtf8(frame.data, 1, descriptionEndIndex);
+      // URL fields are always represented as ISO-8859-1 [ID3v2.4.0 section 4.]
+      // if the value is followed by a string termination all the following information
+      // should be ignored [ID3v2.4.0 section 4.3]
+      frame.url = parseIso88591(
+        frame.data,
+        descriptionEndIndex + 1,
+        frame.data.length
+      ).replace(/\0.*$/, '');
+    },
+    'PRIV': function(frame) {
+      var i;
+
+      for (i = 0; i < frame.data.length; i++) {
+        if (frame.data[i] === 0) {
+          // parse the description and URL fields
+          frame.owner = parseIso88591(frame.data, 0, i);
+          break;
+        }
+      }
+      frame.privateData = frame.data.subarray(i + 1);
+      frame.data = frame.privateData;
+    }
+  };
+
+// TODO add dispatch type parsing
+var parseId3Frames = function(data) {
+  var frameSize, frameHeader,
+    frameStart = 10,
+    tagSize = 0,
+    frames = [];
+
+  // if not valid id3 return early.
+  if (data.length < 10 ||
+    data[0] !== 'I'.charCodeAt(0) ||
+    data[1] !== 'D'.charCodeAt(0) ||
+    data[2] !== '3'.charCodeAt(0)) {
+    return;
+  }
+  // the frame size is transmitted as a 28-bit integer in the
+  // last four bytes of the ID3 header.
+  // The most significant bit of each byte is dropped and the
+  // results concatenated to recover the actual value.
+  tagSize = parseSyncSafeInteger(data.subarray(6, 10));
+
+  // ID3 reports the tag size excluding the header but it's more
+  // convenient for our comparisons to include it
+  tagSize += 10;
+  if (data[5] & 0x40) {
```

Comment:
It may also be useful to assign `data[5]` to a more descriptive variable here.

Reply:
I assume this is a bitmask for flags to check if an extended header is present:
https://github.com/id3/ID3v2.4/blob/master/id3v2.40-structure.txt

Reply:
Yeah, this is the extended header bitmask. I'll add a comment and/or assign this to something more descriptive.